### ISSUE TITLE
AtomGroup.residues

### ISSUE THREAD

**Issue by darmis007** [MDAnalysis/mdanalysis#2687]
## Expected behavior ##
COMMAND - list(u.atoms[100:130].residues)
[<Residue LEU, 6>, <Residue GLY, 7>, <Residue ALA, 8>]

## Actual behavior ##
Traceback (most recent call last):
  File "exmple.py", line 11, in <module>
    print(list(u.atoms.n_residues))
  File "c:\users\user\opens\mdanalysis\package\MDAnalysis\core\groups.py", line 2301, in n_residues
    return len(self.residues)
  File "c:\users\user\opens\mdanalysis\package\MDAnalysis\core\groups.py", line 2261, in residues
    rg = self.universe.residues[unique_int_1d(self.resindices)]
  File "MDAnalysis\lib\_cutil.pyx", line 52, in MDAnalysis.lib._cutil.unique_int_1d
ValueError: Buffer dtype mismatch, expected 'int64_t' but got 'long

The reason for the error is that the unique_int_1d requires a numpy array of int64 datatype while only a simple array is being passed in the function unique_int_1d() so before passing the value it needs to be typed as numpy int64 array

## Code to reproduce the behavior ##

``` python
import MDAnalysis as mda
from MDAnalysis.tests.datafiles import PSF, DCD,  GRO, PDB, TPR, XTC, TRR,  PRMncdf, NCDF

u = mda.Universe(PSF, DCD)
print(list(u.atoms[1:50].residues))


....

```

## Current version of MDAnalysis ##
MDAnalysis version: 0.20.2-dev0
Python version:  Python 3.8.2
Operating System: Windows 10(Home)
- Which version are you using? (run `python -c "import MDAnalysis as mda; print(mda.__version__)"`)
- Which version of Python (`python -V`)?
- Which operating system?

**Comment by orbeckst** [comment]
This works fine on macOS Python 3.7.6
```
[<Residue MET, 1>, <Residue ARG, 2>, <Residue ILE, 3>]
```
I don't have a Windows machine to test locally.

I am not sure if we test this exact problem with our test suite. The first thing for a PR would be to write a test case and see it fail. Then making sure that the types agree. 

@darmis007 Given that you already have identified the potential problem, would you be able to contribute the fix and start a pull request? The [User Guide: Contributing](https://www.mdanalysis.org/UserGuide/contributing.html) section explains in detail how, if you want to know more.

**Comment by darmis007** [comment]
Yes i will contribute to the fix and make a pull request soon

**Comment by lilyminium** [comment]
> I am not sure if we test this exact problem with our test suite. The first thing for a PR would be to write a test case and see it fail. Then making sure that the types agree.


Yes the failing test case would be crucial  @darmis007. The fundamental issue here is that MDAnalysis is relying on C longs being 64-bit, whereas on Windows they are 32-bit. I would assume that this isn’t the only point of failure and we need a way to make sure new code doesn’t do this either. Do the Appveyor Visual Studio architectures not use Windows data types?

**Comment by tylerjereddy** [comment]
> Do the Appveyor Visual Studio architectures not use Windows data types?

They should--probably just need to add an appropriate test. I do have access to a Windows dev machine if we need to check something.

**Comment by richardjgowers** [comment]
Yeah we do have windows builds currently, but they must use a 64 bit
interpretation of long. It’s an oddity of C that long is ill defined.

**Comment by tylerjereddy** [comment]
I was also working on 32-bit support in another PR, may be good to check if this is 32-bit Windows. I don't think devs were necessarily in favor my effort to support 32-bit archs though.

**Comment by IAlibay** [comment]
In terms of why appveyor isn't picking this up, I interestingly cannot reproduce this behaviour on 0.20.1 (not tried dev yet, need to work out how to build it on windows...), which might indicate a type of system dependence.

@darmis007 can you confirm what `System Information` returns for "System Type" (might help narrow down if it's a 32bit thing).

Test system:
OS: Windows 10(Pro) 10.0.18363 Build 18363 
Python: 3.8.2
MDA version: 0.20.1 (conda-forge install)

**Comment by IAlibay** [comment]
Update: a 0.20.2-dev0 build on the same machine as https://github.com/MDAnalysis/mdanalysis/issues/2687#issuecomment-632344623 didn't show any issues.

Note: this is trying to follow as much as possible the appveyor config. However, maybe one other possible avenue for differences is the C++ compiler used. This test used MSC v.1916 64bit.

**Comment by darmis007** [comment]
Sir i have following system config
Windows 10 Home
64bit OS , x64 based processor
MDAnalysis - 0.20.2-dev0

**Comment by IAlibay** [comment]
That's rather interesting, do you know what C++ compiler you are using @darmis007 ?

**Comment by darmis007** [comment]
following gcc compiler
realgcc.exe (Rev1, Built by MSYS2 project) 7.2.0
Copyright (C) 2017 Free Software Foundation, Inc.

**Comment by tylerjereddy** [comment]
@darmis007 Maybe check some of the suggestions here to see if you are using 32-bit Python interpreter: https://stackoverflow.com/questions/1405913/how-do-i-determine-if-my-python-shell-is-executing-in-32bit-or-64bit-mode-on-os

**Comment by tylerjereddy** [comment]
We could add 32-bit interpreter to CI, but I certainly recognize Richard's apprehension about the maintenance burden/value ratio (though it could serve those with access to limited hardware) and whether a low-level (C/Cython) patch is healthier than the type coercion stuff.

**Comment by tylerjereddy** [comment]
For completeness, I've reproduced the reported issue locally in a 32-bit Windows Python interpreter, on a machine that is otherwise "64-bit Windows":

```
Python 3.7.7 (tags/v3.7.7:d7c567b08f, Mar 10 2020, 09:44:33) [MSC v.1900 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
```

```python
>>> import MDAnalysis, MDAnalysisTests
>>> import MDAnalysis as mda
>>> from MDAnalysis.tests.datafiles import PSF, DCD,  GRO, PDB, TPR, XTC, TRR,  PRMncdf, NCDF
>>> u = mda.Universe(PSF, DCD)
>>> print(list(u.atoms[1:50].residues))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\treddy\python37_32bit_venv\lib\site-packages\MDAnalysis\core\groups.py", line 2313, in residues
    rg = self.universe.residues[unique_int_1d(self.resindices)]
  File "MDAnalysis/lib/_cutil.pyx", line 52, in MDAnalysis.lib._cutil.unique_int_1d
ValueError: Buffer dtype mismatch, expected 'int64_t' but got 'long'

```

**Comment by tylerjereddy** [comment]
Do we even have a good reason for `unique_int_1d()` to exist when there's already `np.unique()`? Have our benchmarks really shown that our low-level version is better than theirs?

**Comment by tylerjereddy** [comment]
Maybe we're not providing sorting, but `np.unique()` has keywords for returning indices, so the choice to make a low-level Cython function that is only used in Python in our code base is perhaps worth revisiting.

**Comment by lilyminium** [comment]
`unique_int_1d` does sort the output and is relatively faster; possibly from not needing to check/track the `return_xx` keywords `unique` has, and that its usual input is already sorted

<img width="510" alt="Screenshot 2020-05-24 at 12 14 38 PM" src="https://user-images.githubusercontent.com/31115101/82744208-9d7d4f80-9db8-11ea-9ab1-be72b7fc850d.png">
<img width="514" alt="Screenshot 2020-05-24 at 12 13 59 PM" src="https://user-images.githubusercontent.com/31115101/82744209-a0784000-9db8-11ea-94f9-2b172b9cf361.png">

**Comment by tylerjereddy** [comment]
Alright, I guess I better not substitute it then.

**Comment by tylerjereddy** [comment]
Fixed by #2696, so closing. Please reopen or open a new issue if the issue persists in latest `develop`.
